### PR TITLE
Make index rotation strategies pluggable

### PR DESCRIPTION
**Pull request by bernd**
A first version of making the rotation strategies pluggable. Once we decide that this is the way to go, I will also do the same for the retention strategies.

### Changes

- Change RotationStrategy interface to only require a rotate() method
- Add addRotationStrategy() helper to PluginModule
- Move rotation strategy configuration into the cluster configuration
- Add a ConfigManagementPeriodical which migrates settings from the server configuration file to the database

The TimeBasedRotationStrategy still needs to be a `@Singleton` because it keeps state. This will be moved to the database in a later step.

### Discussion

To add a new rotation strategy, the `addRotationStrategy(MyStrategy.class)` needs to be called in the guice bindings. The strategies take care of getting their configuration on their own. This makes it more flexible than passing in the configuration and also simplifies the bindings and implementation.

As said in the commit message, the `TimeBasedRotationStrategy` still needs to be a `@Singleton` because it keeps state. That state should be moved into the database so the strategy does not need to be a singleton.

### TODO

* The rotation strategies need to be configurable via REST API. This will be added once we decided to go ahead with this.
* The implementation of the rotation strategies can be simplified. I didn't do that yet to keep the changes small and reviewable.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/periodical/IndexRotationThread.java [hunk 1]
```diff
@@ -79,7 +85,19 @@ protected Logger getLogger() {
...
     protected void checkForRotation() {
-        final RotationStrategy rotationStrategy = rotationStrategyProvider.get();
+        final IndexManagementConfig config = clusterConfigService.get(IndexManagementConfig.class);
+
+        if (config == null) {
+            LOG.warn("No index management configuration found, not running index rotation!");
+            return;
+        }
+
+        final RotationStrategy rotationStrategy = rotationStrategyMap.get(config.rotationStrategy()).get();
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/periodical/IndexRotationThread.java]: check for NPE?

File: graylog2-server/src/main/java/org/graylog2/periodical/IndexRotationThread.java [hunk 2]
```diff
@@ -79,7 +85,19 @@ protected Logger getLogger() {
...
+
+        if (rotationStrategy == null) {
+            LOG.warn("No rotation strategy found, not running index rotation!");
+            return;
+        }
 
         String currentTarget;
```
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/periodical/IndexRotationThread.java]: could be moved to abstract superclass?

**Comment by pdepaepe** [comment]
Is this P/R covering this https://github.com/Graylog2/graylog2-server/issues/1555 ?

**Comment by bernd** [comment]
@pdepaepe This is only to make the strategies pluggable to make it possible to implement something like #1555.

**Comment by bernd** [comment]
Rebased after merging Elasticsearch 2.0 branch.

**Comment by joschi** [comment]
Accessing the deflector resource currently generates an internal server error.

Request URL: `http://127.0.0.1:12900/system/deflector/config`
Response Body:
```
{
  "type": "ApiError",
  "message": "A MultiException has 1 exceptions.  They are:\n1. java.lang.reflect.UndeclaredThrowableException\n"
}
```

Exception:
```
015-12-09 16:03:09,179 ERROR: org.graylog2.shared.rest.exceptionmappers.AnyExceptionClassMapper - Unhandled exception in REST resource
org.glassfish.hk2.api.MultiException: A MultiException has 1 exceptions.  They are:
1. java.lang.reflect.UndeclaredThrowableException

	at org.jvnet.hk2.internal.Collector.throwIfErrors(Collector.java:89) ~[hk2-locator-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.ServiceLocatorImpl.secondChanceResolve(ServiceLocatorImpl.java:522) ~[hk2-locator-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetInjecteeDescriptor(ServiceLocatorImpl.java:582) ~[hk2-locator-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getInjecteeDescriptor(ServiceLocatorImpl.java:594) ~[hk2-locator-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.IterableProviderImpl.justInTime(IterableProviderImpl.java:99) ~[hk2-locator-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.IterableProviderImpl.get(IterableProviderImpl.java:108) ~[hk2-locator-2.4.0-b31.jar:?]
	at org.graylog2.rest.resources.system.DeflectorResource.config(DeflectorResource.java:92) ~[classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_66]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_66]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_66]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_66]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102) ~[jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326) [jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271) [jersey-common-2.22.1.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267) [jersey-common-2.22.1.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315) [jersey-common-2.22.1.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297) [jersey-common-2.22.1.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267) [jersey-common-2.22.1.jar:?]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317) [jersey-common-2.22.1.jar:?]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305) [jersey-server-2.22.1.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154) [jersey-server-2.22.1.jar:?]
	at org.graylog2.jersey.container.netty.NettyContainer.messageReceived(NettyContainer.java:356) [jersey-netty-1.5.2.jar:?]
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.10.5.Final.jar:?]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.10.5.Final.jar:?]
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.10.5.Final.jar:?]
	at org.jboss.netty.handler.execution.ChannelUpstreamEventRunnable.doRun(ChannelUpstreamEventRunnable.java:43) [netty-3.10.5.Final.jar:?]
	at org.jboss.netty.handler.execution.ChannelEventRunnable.run(ChannelEventRunnable.java:67) [netty-3.10.5.Final.jar:?]
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:176) [metrics-core-3.1.2.jar:3.1.2]
	at org.jboss.netty.handler.execution.MemoryAwareThreadPoolExecutor$MemoryAwareRunnable.run(MemoryAwareThreadPoolExecutor.java:606) [netty-3.10.5.Final.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_66]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_66]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_66]
Caused by: java.lang.reflect.UndeclaredThrowableException
	at com.sun.proxy.$Proxy152.getBinding(Unknown Source) ~[?:?]
	at org.jvnet.hk2.guice.bridge.internal.GuiceToHk2JITResolver.findBinding(GuiceToHk2JITResolver.java:86) ~[guice-bridge-2.4.0-b31.jar:?]
	at org.jvnet.hk2.guice.bridge.internal.GuiceToHk2JITResolver.justInTimeResolution(GuiceToHk2JITResolver.java:140) ~[guice-bridge-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.ServiceLocatorImpl.secondChanceResolve(ServiceLocatorImpl.java:511) ~[hk2-locator-2.4.0-b31.jar:?]
	... 37 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_66]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_66]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_66]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_66]
	at org.graylog2.shared.bindings.Hk2GuiceBridgeJitInjector.invoke(Hk2GuiceBridgeJitInjector.java:90) ~[classes/:?]
	at com.sun.proxy.$Proxy152.getBinding(Unknown Source) ~[?:?]
	at org.jvnet.hk2.guice.bridge.internal.GuiceToHk2JITResolver.findBinding(GuiceToHk2JITResolver.java:86) ~[guice-bridge-2.4.0-b31.jar:?]
	at org.jvnet.hk2.guice.bridge.internal.GuiceToHk2JITResolver.justInTimeResolution(GuiceToHk2JITResolver.java:140) ~[guice-bridge-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.ServiceLocatorImpl.secondChanceResolve(ServiceLocatorImpl.java:511) ~[hk2-locator-2.4.0-b31.jar:?]
	... 37 more
Caused by: com.google.inject.ConfigurationException: Guice configuration errors:

1) No implementation for org.graylog2.plugin.indexer.rotation.RotationStrategy was bound.
  while locating org.graylog2.plugin.indexer.rotation.RotationStrategy

1 error
	at com.google.inject.internal.InjectorImpl.getBinding(InjectorImpl.java:159) ~[guice-4.0.jar:?]
	at com.google.inject.internal.InjectorImpl.getBinding(InjectorImpl.java:67) ~[guice-4.0.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_66]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_66]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_66]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_66]
	at org.graylog2.shared.bindings.Hk2GuiceBridgeJitInjector.invoke(Hk2GuiceBridgeJitInjector.java:90) ~[classes/:?]
	at com.sun.proxy.$Proxy152.getBinding(Unknown Source) ~[?:?]
	at org.jvnet.hk2.guice.bridge.internal.GuiceToHk2JITResolver.findBinding(GuiceToHk2JITResolver.java:86) ~[guice-bridge-2.4.0-b31.jar:?]
	at org.jvnet.hk2.guice.bridge.internal.GuiceToHk2JITResolver.justInTimeResolution(GuiceToHk2JITResolver.java:140) ~[guice-bridge-2.4.0-b31.jar:?]
	at org.jvnet.hk2.internal.ServiceLocatorImpl.secondChanceResolve(ServiceLocatorImpl.java:511) ~[hk2-locator-2.4.0-b31.jar:?]
	... 37 more
```

**Comment by bernd** [comment]
> Accessing the deflector resource currently generates an internal server error.

Good catch, thank you! I totally missed those.

Updated branch with a fix.

**Comment by bernd** [comment]
Also rebased against master.

**Comment by joschi** [comment]
LGTM. :+1: 

There will have to be follow-up PRs for supporting to change the rotation configuration during runtime.

**Pull request merged**